File: src/main/java/io/github/gdpl2112/DirController.java
package io.github.gdpl2112;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.*;

/**
 * @author github kloping
 * @date 2025/8/24-12:19
 */
@RequestMapping("/dir")
@RestController
public class DirController {
    @Value("${file.upload-dir}")
    private String uploadDir;

    private static final String PARENT_HTML = "<!doctype html>\n" +
            "<html lang=\"zh-CN\">\n" +
            "<head>\n" +
            "    <meta charset=\"UTF-8\">\n" +
            "    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\n" +
            "    <title>KLOPING FILE PORT</title>\n" +
            "    <style>\n" +
            "        body {\n" +
            "            font-family: Arial, sans-serif;\n" +
            "            margin: 20px;\n" +
            "            line-height: 1.6;\n" +
            "        }\n" +
            "        table {\n" +
            "            width: 100%;\n" +
            "            border-collapse: collapse;\n" +
            "            margin-bottom: 20px;\n" +
            "        }\n" +
            "        th, td {\n" +
            "            border: 1px solid #ddd;\n" +
            "            padding: 8px;\n" +
            "            text-align: left;\n" +
            "        }\n" +
            "        th {\n" +
            "            background-color: #f2f2f2;\n" +
            "            font-weight: bold;\n" +
            "        }\n" +
            "        tr:nth-child(even) {\n" +
            "            background-color: #f9f9f9;\n" +
            "        }\n" +
            "        tr:hover {\n" +
            "            background-color: #f1f1f1;\n" +
            "        }\n" +
            "        a.folder {\n" +
            "            color: #8d02ff;\n" +
            "        }\n" +
            "        a.file {\n" +
            "            color: #5a3000;\n" +
            "        }\n" +
            "    </style>\n" +
            "</head>\n" +
            "<body>\n" +
            "<h1>/</h1>\n" +
            "<hr>\n" +
            "<table>\n" +
            "    <thead>\n" +
            "    <tr>\n" +
            "        <th>文件名</th>\n" +
            "        <th>大小</th>\n" +
            "        <th>修改日期</th>\n" +
            "        <th>类型</th>\n" +
            "    </tr>\n" +
            "    </thead>\n" +
            "    <tbody>\n" +
            "    </tbody>\n" +
            "</table>\n" +
            "<hr>\n" +
            "</body>\n" +
            "</html>";

    @RequestMapping
    public Object dir(
            HttpServletRequest request, HttpServletResponse response,
            @RequestParam(value = "path", required = false, defaultValue = "/") String path) {
        // 规范化路径
        String normalizedPath = normalizePath(path);

        File f0 = new File(uploadDir);
        File file = new File(f0, normalizedPath);

        Document document = org.jsoup.Jsoup.parse(PARENT_HTML);
        Element element = document.getElementsByTag("h1").get(0);
        element.text(path);
        List<File> files = new LinkedList<>(Arrays.asList(Objects.requireNonNull(file.listFiles())));
        files.sort((f1,f2)->{
            // 优先文件夹
            if (f1.isDirectory() && !f2.isDirectory()) {
                return -1;
            } else if (!f1.isDirectory() && f2.isDirectory()) {
                return 1;
            } else {
                // 都是文件夹或都是文件，按名称升序
                return f1.getName().compareToIgnoreCase(f2.getName());
            }
        });
        if (!path.equals("/")) {
            Element el0 = new Element("tr");
            Element el1 = new Element("td");
            el1.addClass("folder");

            Element ahref = new Element("a");
            ahref.attr("href", "/dir?path=" + getUpPath(path));
            ahref.text("../");
            el1.appendChild(ahref);
            el0.appendChild(el1);

            document.getElementsByTag("tbody").get(0).appendChild(el0);
        }
        for (File elFile : files) {
            String fname = elFile.getPath();
            fname = fname.replaceAll("\\\\", "/");
            fname = fname.replace(uploadDir, "");
            Element el0 = new Element("tr");

            Element el1 = new Element("td");

            Element ahref = new Element("a");
            ahref.addClass(elFile.isDirectory() ? "folder" : "file");
            String href = elFile.isDirectory() ?
                    "/dir?path=" + fname : fname;
            href = href.replaceAll("//", "/");
            ahref.attr("href", href);
            if (elFile.isFile()) ahref.attr("target", "_blank");
            ahref.text(elFile.getName());
            el1.appendChild(ahref);

            el0.appendChild(el1);

            Element el2 = new Element("td");
            el2.text(getFileSize(elFile));
            el0.appendChild(el2);

            Element el3 = new Element("td");
            el3.text(getTimed(elFile.lastModified()));
            el0.appendChild(el3);

            Element el4 = new Element("td");
            el4.text(elFile.isDirectory() ? "文件夹" : getLastName(elFile.getName()));
            el0.appendChild(el4);
            document.getElementsByTag("tbody").get(0).appendChild(el0);
        }
        String sb = document.html();
        response.setContentType("text/html");
        response.setCharacterEncoding("UTF-8");
        response.setStatus(200);
        try {
            response.getWriter().write(sb);
            response.getWriter().flush();
            response.getWriter().close();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    private static final SimpleDateFormat SF_0 = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss");

    public static String getTimed(long l) {
        return SF_0.format(new Date(l));
    }

    public static String getUpPath(String path) {
        String upPath = path;
        if (upPath.equals("/")) {
            return "/";
        } else {
            int n = upPath.lastIndexOf("/");
            upPath = upPath.substring(0, n);
            return upPath;
        }
    }

    public static String getLastName(String name) {
        int end0 = name.lastIndexOf(".");
        if (end0 < 0) {
            return "未知文件类型";
        }
        return name.substring(end0 + 1) + "文件";
    }

    public static String getFileSize(File listFile) {
        if (listFile.isFile()) {
            return formatBytes(listFile.length());
        } else return "-";
    }

    // 字节单位转换工具（如 B → GB）
    public static String formatBytes(long bytes) {
        if (bytes < 1024) return bytes + " B";
        int exp = (int) (Math.log(bytes) / Math.log(1024));
        char unit = "KMGTPE".charAt(exp - 1);
        return String.format("%.2f %sB", bytes / Math.pow(1024, exp), unit);
    }

    /**
     * 规范化路径，防止路径遍历攻击
     */
    public static String normalizePath(String path) {
        if (path == null || path.isEmpty()) {
            return "/";
        }

        // 替换反斜杠为正斜杠
        path = path.replace('\\', '/');

        // 处理相对路径
        String[] parts = path.split("/");
        List<String> normalizedParts = getNormalizedParts(parts);

        // 重新组合路径
        StringBuilder normalizedPath = new StringBuilder();
        for (String part : normalizedParts) {
            normalizedPath.append("/").append(part);
        }

        return normalizedPath.length() > 0 ? normalizedPath.toString() : "/";
    }

    private static List<String> getNormalizedParts(String[] parts) {
        List<String> normalizedParts = new ArrayList<>();

        for (String part : parts) {
            if (part.equals("..")) {
                // 回退上级目录，但不能超出根目录
                if (!normalizedParts.isEmpty() && !normalizedParts.get(normalizedParts.size() - 1).equals("..")) {
                    normalizedParts.remove(normalizedParts.size() - 1);
                } else {
                    // 如果试图超出根目录，则忽略
                    continue;
                }
            } else if (!part.equals(".") && !part.isEmpty()) {
                normalizedParts.add(part);
            }
        }
        return normalizedParts;
    }
}

File: gofiles/controller/DirController.go
package controller

import (
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/gin-gonic/gin"
)

// DirController 目录控制器
type DirController struct {
	UploadDir string
}

// NewDirController 创建目录控制器实例
func NewDirController(uploadDir string) *DirController {
	return &DirController{
		UploadDir: uploadDir,
	}
}

// Register 注册路由
func (c *DirController) Register(router *gin.Engine) {
	router.GET("/dir", c.Dir)
}

// Dir 目录浏览
func (c *DirController) Dir(ctx *gin.Context) {
	path := ctx.DefaultQuery("path", "/")
	normalizedPath := c.normalizePath(path)

	filePath := filepath.Join(c.UploadDir, normalizedPath)
	fileInfo, err := os.Stat(filePath)
	if err != nil || !fileInfo.IsDir() {
		ctx.JSON(http.StatusNotFound, gin.H{"message": "目录不存在"})
		return
	}

	entries, err := os.ReadDir(filePath)
	if err != nil {
		ctx.JSON(http.StatusInternalServerError, gin.H{"message": "读取目录失败"})
		return
	}

	// 排序：文件夹优先，然后按名称排序
	sort.Slice(entries, func(i, j int) bool {
		iIsDir := entries[i].IsDir()
		jIsDir := entries[j].IsDir()

		if iIsDir && !jIsDir {
			return true
		}
		if !iIsDir && jIsDir {
			return false
		}

		return strings.ToLower(entries[i].Name()) < strings.ToLower(entries[j].Name())
	})

	// 生成HTML响应
	html := c.generateDirHTML(path, entries, normalizedPath)
	ctx.Header("Content-Type", "text/html; charset=utf-8")
	ctx.String(http.StatusOK, html)
}

// normalizePath 规范化路径
func (c *DirController) normalizePath(path string) string {
	path = strings.ReplaceAll(path, "\\", "/")
	parts := strings.Split(path, "/")
	normalizedParts := make([]string, 0)

	for _, part := range parts {
		if part == ".." {
			if len(normalizedParts) > 0 && normalizedParts[len(normalizedParts)-1] != ".." {
				normalizedParts = normalizedParts[:len(normalizedParts)-1]
			}
		} else if part != "." && part != "" {
			normalizedParts = append(normalizedParts, part)
		}
	}

	return strings.Join(normalizedParts, "/")
}

// getUpPath 获取上级目录路径
func (c *DirController) getUpPath(path string) string {
	if path == "/" {
		return "/"
	}

	lastSlash := strings.LastIndex(path, "/")
	if lastSlash == -1 {
		return "/"
	}

	return path[:lastSlash]
}

// getFileSize 获取文件大小
func (c *DirController) getFileSize(filePath string) string {
	fileInfo, err := os.Stat(filePath)
	if err != nil {
		return "-"
	}

	if fileInfo.IsDir() {
		return "-"
	}

	return c.formatBytes(fileInfo.Size())
}

// formatBytes 格式化字节大小
func (c *DirController) formatBytes(bytes int64) string {
	if bytes < 1024 {
		return fmt.Sprintf("%d B", bytes)
	}

	units := []string{"KB", "MB", "GB", "TB"}
	size := float64(bytes) / 1024
	unitIndex := 0

	for size >= 1024 && unitIndex < len(units)-1 {
		size /= 1024
		unitIndex++
	}

	return fmt.Sprintf("%.2f %s", size, units[unitIndex])
}

// getFileExtension 获取文件扩展名
func (c *DirController) getFileExtension(filename string) string {
	ext := filepath.Ext(filename)
	if ext == "" {
		return "未知文件类型"
	}
	return strings.TrimPrefix(ext, ".") + "文件"
}

// generateDirHTML 生成目录浏览HTML
func (c *DirController) generateDirHTML(path string, entries []os.DirEntry, normalizedPath string) string {
	var html strings.Builder

	html.WriteString(`<!doctype html>
<html lang="zh-CN">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>KLOPING FILE PORT</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            margin: 20px;
            line-height: 1.6;
        }
        table {
            width: 100%;
            border-collapse: collapse;
            margin-bottom: 20px;
        }
        th, td {
            border: 1px solid #ddd;
            padding: 8px;
            text-align: left;
        }
        th {
            background-color: #f2f2f2;
            font-weight: bold;
        }
        tr:nth-child(even) {
            background-color: #f9f9f9;
        }
        tr:hover {
            background-color: #f1f1f1;
        }
        a.folder {
            color: #8d02ff;
        }
        a.file {
            color: #5a3000;
        }
    </style>
</head>
<body>
`)

	html.WriteString(fmt.Sprintf("<h1>%s</h1>\n", path))
	html.WriteString("<hr>\n")
	html.WriteString(`<table>
    <thead>
    <tr>
        <th>文件名</th>
        <th>大小</th>
        <th>修改日期</th>
        <th>类型</th>
    </tr>
    </thead>
    <tbody>
`)

	// 添加上级目录链接
	if path != "/" {
		upPath := c.getUpPath(path)
		html.WriteString(fmt.Sprintf(`    <tr>
        <td><a href="/dir?path=%s" class="folder">../</a></td>
        <td>-</td>
        <td>-</td>
        <td>文件夹</td>
    </tr>
`, upPath))
	}

	// 添加文件和文件夹列表
	for _, entry := range entries {
		entryPath := filepath.Join(normalizedPath, entry.Name())
		webPath := strings.ReplaceAll(entryPath, "\\", "/")

		var href, cssClass, fileType string
		if entry.IsDir() {
			href = fmt.Sprintf("/dir?path=%s", webPath)
			cssClass = "folder"
			fileType = "文件夹"
		} else {
			href = "/" + webPath
			cssClass = "file"
			fileType = c.getFileExtension(entry.Name())
		}

		fileInfo, _ := entry.Info()
		modTime := fileInfo.ModTime().Format("2006/01/02 15:04:05")
		fileSize := c.getFileSize(filepath.Join(c.UploadDir, entryPath))

		targetBlank := ""
		if !entry.IsDir() {
			targetBlank = ` target="_blank"`
		}
		html.WriteString(fmt.Sprintf(`    <tr>
        <td><a href="%s" class="%s"%s>%s</a></td>
        <td>%s</td>
        <td>%s</td>
        <td>%s</td>
    </tr>
`,
			href, cssClass, targetBlank,
			entry.Name(), fileSize, modTime, fileType))
	}

	html.WriteString(`    </tbody>
</table>
<hr>
</body>
</html>`)

	return html.String()
}
